source/Halibut.Tests/DataStreamFixture.cs

```diff
@@ -0,0 +1,104 @@
+                {
+                    // Note that the DataStream is only given a sync writer.
+                    var count = await echo.CountBytesAsync(new DataStream(data.Length, stream => stream.Write(data, 0, data.Length)));
+                    count.Should().Be(data.Length);
+                }
+            }
+        }
+        
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false, testAsyncAndSyncClients: false)]
+        public async Task AsyncDataStreamsAreUsedWhenInAsync(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            using (var clientAndService = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                       .WithStandardServices()
+                       .AsLatestClientAndLatestServiceBuilder()
+                       .WithForcingClientProxyType(ForceClientProxyType.AsyncClient)
+                       .Build(CancellationToken))
+            {
+                var echo = clientAndService.Client.CreateAsyncClient<IEchoService, IAsyncClientEchoService>(clientAndService.GetServiceEndPoint());
+
+                var data = new byte[1337];
+                new Random().NextBytes(data);
+
+                for (var i = 0; i < clientAndServiceTestCase.RecommendedIterations; i++)
+                {
+                    var count = await echo.CountBytesAsync(new DataStream(data.Length, 
+                        _ => throw new Exception("You have no power here Gandalf the sync"), 
+                        async (stream, ct) => await stream.WriteByteArrayAsync(data, ct)
+                        )
+                    );
+                    count.Should().Be(data.Length);
+                }
+            }
+        }
+        
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testNetworkConditions: false)]
+        public async Task AsyncDataStreamsCanBeUsedInSyncAndAsync(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            using (var clientAndService = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                       .WithStandardServices()
+                       .Build(CancellationToken))
+            {
+                var echo = clientAndService.CreateClient<IEchoService, IAsyncClientEchoService>();
+
+                var data = new byte[1337];
+                new Random().NextBytes(data);
+
+                for (var i = 0; i < clientAndServiceTestCase.RecommendedIterations; i++)
+                {
+                    // Note this only has a async writer.
+                    var count = await echo.CountBytesAsync(new DataStream(data.Length,
+                            async (stream, ct) => await stream.WriteByteArrayAsync(data, ct))
+                    );
+                    count.Should().Be(data.Length);
+                }
+            }
+        }
+
+        [Test]
+        public async Task ASyncDataStreamWriter_CanBeUsedInAsync()
```

Comment:
The other tests in this class are more "integration style" tests, whereas this is more a "unit style" test.

Could we exercise this in an integration style way?

If not, I wonder if it would be clearer to separate this test from the others (in a new class). Personally, I like having all tests in a class being tested at a similar level 😁